src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php

```diff
@@ -24,6 +25,7 @@ class DateTimeNormalizer implements NormalizerInterface, DenormalizerInterface,
 {
     const FORMAT_KEY = 'datetime_format';
     const TIMEZONE_KEY = 'datetime_timezone';
+    const PRESERVE_CONTEXT_TIMEZONE = 'preserve_context_timezone';
```

Comment:
I think reversing the logic with a `PRESERVE_INPUT_TIMEZONE` will be easier to reason about, as the default behavior will be to always use the context timezone in the future.
So it seems more natural to me to set a `PRESERVE_INPUT_TIMEZONE` to `true` rather than setting a `PRESERVE_CONTEXT_TIMEZONE` to false.

Also, it clearly mentions this is about the input, so only about denormalization.

Reply:
Why wouldn't we apply the same logic when normalizing?

> I feel this would again muddy the water here: if I'm explicitly specifying a timezone for the normalizer, why wouldn't it apply it when normalizing? The \Datetime instance could have arrived from many sources so it might be a chore on the developer to set it manually everywhere (maybe even recursively) before serialization.

Reply:
It makes no sense when talking about normalization: we already always use the timezone from the `TIMEZONE_KEY` to output dates with consistent timezones.

> if I'm explicitly specifying a timezone for the normalizer, why wouldn't it apply it when normalizing?

This is the case (and would still be), right? Or is there any misunderstanding on one side?

Reply:
Setting the flag to true by default is BC break.  Hence, set to false preserves the current behaviour if the the user does nothing and Symfony < 5.0, but issues a deprecation in the key is missing below Symfony 5.0  Above 5.0 the flag is ignored and assumed to be true all the time.

Reply:
- PRESERVE_CONTEXT_TIMEZONE is a temporary key from NOW to version 5.0, where it is no longer required.
- Current code does **NOT** preserve the context timezone on denormalisation, but does on normalisation.  This could be considered the bug, but some see it as correct behaviour.
- From now to 5.0:
    - not setting the PRESERVE_CONTEXT_TIMEZONE flag maintains the current behaviour. (denormalisation can ignore context timezone.) This preserves BC from now to 5.0.
    - setting a context timezone with setting PRESERVE_CONTEXT_TIMEZONE throws a deprecation to alert user to the upcoming change of behaviour, and the denormaliser uses PHP default behaviour.
   - the PRESERVE_CONTEXT_TIMEZONE set to false to remove the deprecation warning, and the denormaliser uses PHP default behaviour.
    - the PRESERVE_CONTEXT_TIMEZONE set to true to remove the deprecation warning, and denormalisation always applies the context timezone, unless context timezone is null.
- after 5.0:
    - this is BC break.
    - the PRESERVE_CONTEXT_TIMEZONE flag is ignored.
    - Denormalisation always applies the context timezone.
    - defaults to PHP behaviour if context timezone is null.